src/main/scala/akka/contrib/process/BlockingProcess.scala

```diff
@@ -107,29 +122,53 @@ class BlockingProcess(
     }
 
     try {
-      val stdin =
-        context.actorOf(OutputStreamSubscriber.props(process.getOutputStream), "stdin")
+      val selfRef = context.self
 
-      val stdout =
-        context.watch(context.actorOf(InputStreamPublisher.props(process.getInputStream, stdioTimeout), "stdout"))
+      val stdin = Sink.outputStream(process.getOutputStream)
+        .mapMaterializedValue(_.andThen { case _ => selfRef ! StdinTerminated })
 
-      val stderr =
-        context.watch(context.actorOf(InputStreamPublisher.props(process.getErrorStream, stdioTimeout), "stderr"))
+      val stdout = Source.inputStream(process.getInputStream)
+        .withAttributes(Attributes.inputBuffer(1, 1))
+        .mapMaterializedValue(_.andThen { case _ => selfRef ! StdoutTerminated })
 
-      context.parent ! Started(ActorSubscriber(stdin), ActorPublisher(stdout), ActorPublisher(stderr))
+      val stderr = Source.inputStream(process.getErrorStream)
+        .withAttributes(Attributes.inputBuffer(1, 1))
+        .mapMaterializedValue(_.andThen { case _ => selfRef ! StderrTerminated })
+
+      context.parent ! Started(stdin, stdout, stderr)
     } finally {
       context.watch(context.actorOf(ProcessDestroyer.props(process, context.parent), "process-destroyer"))
     }
   }
 
-  override def receive = {
+  override def receive = handleMessages(stdOutTerminated = false, stdErrTerminated = false)
+
+  private def handleMessages(stdOutTerminated: Boolean, stdErrTerminated: Boolean): Receive = {
     case Destroy =>
       log.debug("Received request to destroy the process.")
       context.stop(self)
     case Terminated(ref) =>
       log.debug("Child {} was unexpectedly stopped, shutting down process", ref.path)
       context.stop(self)
+    case StdinTerminated =>
+      log.debug("Stdin was terminated")
+    case StdoutTerminated =>
+      log.debug("Stdout was terminated")
+      context.become({
+        case StderrTerminated =>
+          log.debug("Stdout and Stderr was both terminated - shutting down process")
+          context.stop(self)
+      }, discardOld = false)
```

Comment:
But if we don't have `discardOld = false` it won't respond to `Destroy` or `Terminated`